session.go

```diff
@@ -919,15 +914,77 @@ func (s *session) DropPreparedStmt(stmtID uint32) error {
 	return nil
 }
 
-// StmtTxn wraps kv.Transaction to provide a new kv.Transaction.
-// It holds all statement related modification in the buffer before flush to the txn,
+// TxnState wraps kv.Transaction to provide a new kv.Transaction.
+// 1. It holds all statement related modification in the buffer before flush to the txn,
 // so if execute statement meets error, the txn won't be made dirty.
-type StmtTxn struct {
+// 2. It's a lazy transaction, that means it's a txnFuture befort StartTS() is really need.
+type TxnState struct {
 	kv.Transaction
+	txnFuture *txnFuture
 
 	buf          kv.MemBuffer
 	mutations    map[int64]*binlog.TableMutation
 	dirtyTableOP []dirtyTableOperation
+
+	// If StmtCommit meets error (which should not happen, just in case), mark it.
+	// And rollback the whole transaction when it commit.
+	fail error
+}
+
+func (st *TxnState) init() {
+	st.buf = kv.NewMemDbBuffer(kv.DefaultTxnMembufCap)
+	st.mutations = make(map[int64]*binlog.TableMutation)
+}
+
+// Valid overrides Transaction interface.
+func (st *TxnState) Valid() bool {
+	return st.Transaction != nil && st.Transaction.Valid()
+}
+
+func (st *TxnState) validOrPending() bool {
+	return st.txnFuture != nil || st.Valid()
+}
+
+func (st *TxnState) String() string {
+	if st.Transaction != nil {
+		return st.Transaction.String()
+	}
+	if st.txnFuture != nil {
+		return "txnFuture"
+	}
+	return "invalid transaction"
+}
+
+func (st *TxnState) changeInvalidToValid(txn kv.Transaction) {
+	st.Transaction = txn
+	st.txnFuture = nil
+}
+
+func (st *TxnState) changeInvalidToPending(future *txnFuture) {
+	st.Transaction = nil
+	st.txnFuture = future
+}
+
+func (st *TxnState) changePendingToValid() error {
+	if st.txnFuture == nil {
+		return errors.New("transaction future is not set")
```

Comment:
What should the user do when he meet this error?

Reply:
The code is moved from here:
https://github.com/pingcap/tidb/pull/5763/files#diff-f2fccfb96064097b73fbe8eeae3703eeL1501

If somebody make the wrong logic, i.e. never PrepareTxn but call ActivePendingTxn, the code could go to this line.
I think panic is okay when the caller meet this...because it should never happen.